src/linker/Linker/MessageOrigin.cs

```diff
@@ -20,29 +23,48 @@ public MessageOrigin (string fileName, int sourceLine = 0, int sourceColumn = 0)
 			FileName = fileName;
 			SourceLine = sourceLine;
 			SourceColumn = sourceColumn;
+			MdTokenProvider = null;
 		}
 
-		public static MessageOrigin? TryGetOrigin (IMemberDefinition sourceMethod, int ilOffset)
+		public MessageOrigin (IMetadataTokenProvider mdTokenProvider, int sourceLine = 0, int sourceColumn = 0)
+		{
+			FileName = null;
+			MdTokenProvider = mdTokenProvider;
+			SourceLine = sourceLine;
+			SourceColumn = sourceColumn;
+		}
+
+		private MessageOrigin (string fileName, IMetadataTokenProvider mdTokenProvider, int sourceLine = 0, int sourceColumn = 0)
+		{
+			FileName = fileName;
+			MdTokenProvider = mdTokenProvider;
+			SourceLine = sourceLine;
+			SourceColumn = sourceColumn;
+		}
+
+		public static MessageOrigin TryGetOrigin (IMetadataTokenProvider sourceMethod, int ilOffset)
```

Comment:
This is weird - if I have an IL offset then I must have a method, and if I don't have a method I can't have IL offset. So I think this should take the first parameter as `MethodDefinition`.